### PR TITLE
Add ondemand::value.type()

### PR DESCRIPTION
**Pull request by jkeiser**
This adds `type()` to ondemand::value, making it easier to write recursive DOM generators with On Demand.

Also adds `value_unsafe()` to ondemand's unfortunately separate simdjson_result_base.

### CODE REVIEW

**Comment by jkeiser** [comment]
OK, this triggered target mismatch error again when I tried to add json_type to the top level. I'll make a proposal for pre-1.0, I think we need to fix the inlining problem sooner than later or something weird is going to go wrong somewhere. At the least, it's going to be harder to modify simdjson :)

File: include/simdjson/generic/ondemand/document.h [hunk 1]
```diff
@@ -263,6 +263,17 @@ class document {
   /** @overload simdjson_really_inline simdjson_result<value> find_field_unordered(std::string_view key) & noexcept; */
   simdjson_really_inline simdjson_result<value> operator[](const char *key) & noexcept;
 
+  /**
+   * Get the type of this JSON value.
```
Comment (lemire) [inline include/simdjson/generic/ondemand/document.h:267]: I would list the possible values: json_type::array,  json_type::object, json_type::number, json_type::string,  json_type::boolean, json_type::null.

(Optional comment.)

**Review by lemire — APPROVED** [review]
This is very good.

It deviates somewhat from the DOM API as far as how numbers are identified. So you don't know whether you are dealing with, say, a 64-bit unsigned int or a double. What you did is very good and should not be changed... I am just making the remark.

File: include/simdjson/generic/ondemand/document.h (hunk 1 above)
Comment (jkeiser) [reply include/simdjson/generic/ondemand/document.h:267]: That's a good idea here. I'll do that.

**Comment by jkeiser** [comment]
Yep, I agree, matching the JSON type is the right thing in this case (since we're dealing with raw tokens rather than converted values).

**Pull request merged**